example/hyperparam/train_keras.py

```diff
@@ -0,0 +1,143 @@
+
+    def on_epoch_end(self, epoch, logs=None):
+        if not logs:
+            return
+        v = logs["val_loss"]
+        if v < self._best:
+            # The result improved in the validation set.
+            # Log the model with mlflow and also evaluate and log on test set.
+            self._best = v
+            preds = self.model.predict((self._test_x))
+            rmse, mae, r2 = eval_metrics(self._test_y, preds)
+            print("  RMSE: %s" % rmse)
+            print("  MAE: %s" % mae)
+            print("  R2: %s" % r2)
+            mlflow.log_metric("rmse", rmse)
+            mlflow.log_metric("mae", mae)
+            mlflow.log_metric("r2", r2)
+            mlflow.keras.log_model(self.model, "model_checkpoint_{}".format(epoch + 1))
+            mlflow.keras.log_model(self.model, "model")
+
+
+@click.command(help="Trains an Keras model on wine-quality dataset."
+                    "The input is expected in csv format."
+                    "The model and its metrics are logged with mlflow.")
+@click.option("--epochs", type=click.INT, default=100, help="Maximum number of epochs to evaluate.")
+@click.option("--batch-size", type=click.INT, default=16,
+              help="Batch size passed to the learning algo.")
+@click.option("--learning-rate", type=click.FLOAT, default=1e-2, help="Learning rate.")
+@click.option("--beta1", type=click.FLOAT, default=.9, help="Learning rate decay.")
+@click.option("--beta2", type=click.FLOAT, default=.9, help="Gradient moving average decay.")
+@click.option("--seed", type=click.INT, default=97531, help="Seed for the random generator.")
+@click.argument("training_data")
+def run(training_data, epochs, batch_size, learning_rate, beta1, beta2, seed):
+    warnings.filterwarnings("ignore")
+    # Read the wine-quality csv file (make sure you're running this from the root of MLflow!)
+    data = pd.read_csv(training_data)
+    # Split the data into training and test sets. (0.75, 0.25) split.
+    train, test = train_test_split(data, random_state=seed)
+    train, valid = train_test_split(train, random_state=seed)
+    # The predicted column is "quality" which is a scalar from [3, 9]
+    train_x = train.drop(["quality"], axis=1).as_matrix()
+    train_x = (train_x).astype('float32')
+    train_y = train[["quality"]].as_matrix().astype('float32')
+    valid_x = (valid.drop(["quality"], axis=1).as_matrix()).astype('float32')
+
+    valid_y = valid[["quality"]].as_matrix().astype('float32')
+
+    test_x = (test.drop(["quality"], axis=1).as_matrix()).astype("float32")
+    test_y = test[["quality"]].as_matrix().astype("float32")
+
+    with mlflow.start_run():
+        rmse, mae, r2 = eval_metrics(test_y, np.ones(len(test_y)) * np.mean(test_y))
+        mlflow.log_metric("rmse_null", rmse)
+        mlflow.log_metric("mae_null", mae)
+        mlflow.log_metric("r2_null", r2)
+        print("  RMSE (null): %s" % rmse)
+        print("  MAE (null) : %s" % mae)
+        print("  R2 (null): %s" % r2)
```

Comment:
Why do we want to log these separately?

Reply:
Good question. 2 reasons:

So that the caller can use the null model metric as max reasonable value. This is needed in the GpyOpt because some configurations fail or return NaNs or Inf. In that case I need to be able to fill in reasonable value (math.inf does not work unfortunately) which is >= all other values.

I think it is always useful to compare the metric against the value obtained from the null model.

I could always log the metric of the null model as the first value of the metric, but then the user might need to get the history and take the first value which seems more error prone.